File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/presentation/paths/tasks/TasksDiffCallback.java
package com.zhuinden.simplestackdemoexamplefragments.presentation.paths.tasks;

import android.support.v7.util.DiffUtil;

import com.zhuinden.simplestackdemoexamplefragments.presentation.objects.Task;

import java.util.List;

/**
 * Created by Zhuinden on 2017.01.27..
 */

class TasksDiffCallback
        extends DiffUtil.Callback {
    private List<Task> oldTasks;
    private List<Task> newTasks;

    public TasksDiffCallback(List<Task> oldTasks, List<Task> newTasks) {
        this.oldTasks = oldTasks;
        this.newTasks = newTasks;
    }

    @Override
    public int getOldListSize() {
        return oldTasks == null ? 0 : oldTasks.size();
    }

    @Override
    public int getNewListSize() {
        return newTasks.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return newTasks.get(newItemPosition).getId().equals(oldTasks.get(oldItemPosition).getId());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return newTasks.get(newItemPosition).equals(oldTasks.get(oldItemPosition));
    }
}

File: examples/legacy-architecture-samples/simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/features/tasks/TasksDiffCallback.kt
package com.zhuinden.simplestackdemoexamplemvp.features.tasks

import android.support.v7.util.DiffUtil

import com.zhuinden.simplestackdemoexamplemvp.domain.Task

/**
 * Created by Zhuinden on 2018. 08. 20.
 */

class TasksDiffCallback(
    private val oldTasks: List<Task>?,
    private val newTasks: List<Task>
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = oldTasks?.size ?: 0

    override fun getNewListSize(): Int = newTasks.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
        newTasks[newItemPosition].id == oldTasks!![oldItemPosition].id

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
        newTasks[newItemPosition] == oldTasks!![oldItemPosition]
}